tests/libp2p/test_notify.py

```diff
@@ -0,0 +1,152 @@
+"""
+Test Notify and Notifee by ensuring that the proper events get
+called, and that the stream passed into opened_stream is correct
+
+Note: Listen event does not get hit because MyNotifee is passed
+into network after network has already started listening
+
+TODO: Add tests for closed_stream disconnected, listen_close when those
+features are implemented in swarm
+"""
+
+import pytest
+
+from tests.utils import cleanup
+from libp2p import new_node
+from libp2p.network.notifee_interface import INotifee
+
+# pylint: disable=too-many-locals
+
+class MyNotifee(INotifee):
+    # pylint: disable=too-many-instance-attributes, cell-var-from-loop
+
+    def __init__(self, events, val_to_append_to_event):
+        self.events = events
+        self.val_to_append_to_event = val_to_append_to_event
+
+    async def opened_stream(self, network, stream):
+        self.events.append(["opened_stream" + \
+            self.val_to_append_to_event, stream])
+
+    async def closed_stream(self, network, stream):
+        pass
+
+    async def connected(self, network, conn):
+        self.events.append(["connected" + self.val_to_append_to_event,\
+            conn])
+
+    async def disconnected(self, network, conn):
+        pass
+
+    async def listen(self, network, multiaddr):
+        pass
+
+    async def listen_close(self, network, multiaddr):
+        pass
+
+async def perform_two_host_simple_set_up():
+    node_a = await new_node(transport_opt=["/ip4/127.0.0.1/tcp/0"])
+    node_b = await new_node(transport_opt=["/ip4/127.0.0.1/tcp/0"])
+
+    async def my_stream_handler(stream):
+        while True:
+            read_string = (await stream.read()).decode()
+
+            resp = "ack:" + read_string
+            await stream.write(resp.encode())
+
+    node_b.set_stream_handler("/echo/1.0.0", my_stream_handler)
+
+    # Associate the peer with local ip address (see default parameters of Libp2p())
+    node_a.get_peerstore().add_addrs(node_b.get_id(), node_b.get_addrs(), 10)
+    return node_a, node_b
+
+@pytest.mark.asyncio
+async def test_one_notifier():
+    node_a, node_b = await perform_two_host_simple_set_up()
+
+    # Add notifee for node_a
```

Comment:
@robzajac Good call on testing notifees for node_b as well. Will get on that. For the `listen` event, given our current architecture, swarm starts listening when the libp2p host gets created and before the host gets returned. So, swarm starts listening before `Notify` can be called on swarm. So we would probably need to restructure libp2p host creation (or allow for swarm to be passed in) in order to use the `listen` event. See libp2p host creation [here](https://github.com/libp2p/py-libp2p/blob/master/libp2p/__init__.py#L42)

Reply:
> So we would probably need to restructure libp2p host creation (or allow for swarm to be passed in) in order to use the listen event.

This would also be beneficial for this PR https://github.com/libp2p/py-libp2p/pull/129. Right now since the swarm starts listening during host creation, I cannot make the dht start listening at the same time, which then requires to call `dht.listen` later on which is a bit cumbersome.